### PR TITLE
Experimental api support for HostOutputs

### PR DESCRIPTION
**Pull request by chsienki**
This PR introduces the notion of the 'Experimental' API surface, via `Microsoft.CodeAnalysis.Experimental`. That assembly has IVT into `Microsoft.CodeAnalysis` and can 'promote' internal APIs to become public. 

Here' we're creating an internal API for host outputs and exposing it via the new experimental package. 

I suspect this approach may be controversial, and to be clear, we haven't decided on this being the path forward at all, but it seemed the best way to think about it was to implement it and get feedback.

### CODE REVIEW

**Comment by chsienki** [comment]
image

**Comment by CyrusNajmabadi** [comment]
I'm personally a fan.  Would also love it if someone using these had to opt-in somehow as well.  e.g. an addition to their project file like `<AllowExperimentalRoslynBitsThatMayCHangeAtAnyMomentAndBreakEverythingDownstreamThatDependsOnItYesIUnderstandTheImplicationsHereKThxBye/>`

**Comment by chsienki** [comment]
@CyrusNajmabadi The main idea was that you had to explicitly add a package ref to `Microsoft.CodeAnalysis.Experimental` which acts as the opt-in. I.e. you don't get these by default (and we can see the ref in things like binlogs too, which might be useful)

But happy to add an extra swtich on top of that if we want to make it *Really* clear :)

File: src/Compilers/Core/Experimental/Microsoft.CodeAnalysis.Experimental.csproj [hunk 1]
```diff
@@ -0,0 +1,18 @@
+﻿<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <TargetFramework>netstandard2.0</TargetFramework>
```
Comment (jaredpar) [inline src/Compilers/Core/Experimental/Microsoft.CodeAnalysis.Experimental.csproj:4]: For simplicity we should have same TF here as we do in MS.CA.

File: src/Compilers/Core/Portable/SourceGeneration/IncrementalContexts.cs [hunk 2]
```diff
@@ -166,16 +167,19 @@ internal SourceProductionContext(AdditionalSourcesCollection sources, Diagnostic
 
         internal readonly GeneratorRunStateTable.Builder GeneratorRunStateBuilder;
 
+        internal readonly ArrayBuilder<(string, string)> HostOutputBuilder;
```
Comment (jaredpar) [inline src/Compilers/Core/Portable/SourceGeneration/IncrementalContexts.cs:170]: Please add names to the tuple elements here.

File: Compilers.sln [hunk 3]
```diff
@@ -162,6 +162,8 @@ Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "vbc-arm64", "src\Compilers\
 EndProject
 Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Microsoft.CodeAnalysis.CSharp.EndToEnd.UnitTests", "src\Compilers\CSharp\Test\EndToEnd\Microsoft.CodeAnalysis.CSharp.EndToEnd.UnitTests.csproj", "{F3D9264A-7CAE-4265-AF48-0C863301F51E}"
 EndProject
+Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "Microsoft.CodeAnalysis.Experimental", "src\Compilers\Core\Experimental\Microsoft.CodeAnalysis.Experimental\Microsoft.CodeAnalysis.Experimental.csproj", "{62945D52-AB03-4403-97C3-EF39D64FDE77}"
```
Comment (jaredpar) [inline Compilers.sln:165]: This is going to add an image load to Razor scenarios and at some ponit Rsolyn ones too. We should pre-clear this with the VS perf team.

**Pull request closed without merging**